File: xenon-android-library/src/main/kotlin/com/abubusoft/xenon/animations/AbstractAnimationManager.java
package com.abubusoft.xenon.animations;

import java.util.ArrayList;
import java.util.HashMap;

@SuppressWarnings("rawtypes")
public abstract class AbstractAnimationManager<T extends Timeline, A extends Animation> {

	protected AbstractAnimationManager()
	{
		timelineMap = new HashMap<>();
		timelineList = new ArrayList<>();
		animationMap = new HashMap<>();
		animationList = new ArrayList<>();
	}
	
	public abstract void clear();
	
	/**
	 * lista delle animazioni
	 */
	protected ArrayList<A> animationList;

	/**
	 * mappa delle animazioni
	 */
	protected HashMap<String, A> animationMap;

	/**
	 * lista delle timeline
	 */
	protected ArrayList<T> timelineList;

	/**
	 * mappa delle timeline
	 */
	protected HashMap<String, T> timelineMap;
	
	/**
	 * Recupera un timeline in base al suo index.
	 * 
	 * @param index
	 * @return
	 */
	public T getTimeline(int index) {
		return timelineList.get(index);
	}

	/**
	 * Recupera un timeline in base al suo nome.
	 * 
	 * @param animatorId
	 * @return
	 */
	public T getTimeline(String animatorId) {
		return timelineMap.get(animatorId);
	}

	public A getAnimation(String animationName) {
		return animationMap.get(animationName);
	}

	public A getAnimation(int index) {
		return animationList.get(index);
	}
	
	public int getTimelineCount() {
		return timelineList.size();
	}

	public int getAnimationCount() {
		return animationList.size();
	}
}

File: xenon-android-library/src/main/kotlin/com/abubusoft/xenon/animations/AbstractAnimationManager.kt
package com.abubusoft.xenon.animations

abstract class AbstractAnimationManager<T : Timeline<*, *, *>?, A : Animation<*>?> protected constructor() {
    abstract fun clear()

    /**
     * lista delle animazioni
     */
    protected var animationList: ArrayList<A>

    /**
     * mappa delle animazioni
     */
    protected var animationMap: HashMap<String, A>

    /**
     * lista delle timeline
     */
    protected var timelineList: ArrayList<T>

    /**
     * mappa delle timeline
     */
    protected var timelineMap: HashMap<String, T>

    init {
        timelineMap = HashMap()
        timelineList = ArrayList()
        animationMap = HashMap()
        animationList = ArrayList()
    }

    /**
     * Recupera un timeline in base al suo index.
     *
     * @param index
     * @return
     */
    fun getTimeline(index: Int): T {
        return timelineList[index]
    }

    /**
     * Recupera un timeline in base al suo nome.
     *
     * @param animatorId
     * @return
     */
    fun getTimeline(animatorId: String): T? {
        return timelineMap[animatorId]
    }

    fun getAnimation(animationName: String): A? {
        return animationMap[animationName]
    }

    fun getAnimation(index: Int): A {
        return animationList[index]
    }

    val timelineCount: Int
        get() = timelineList.size
    val animationCount: Int
        get() = animationList.size
}